### PR TITLE
[example] Create EEG-GCNN example.

### PR DESCRIPTION
**Pull request by JOHNW02**
## Description
This is an EEG-GCNN example using DGL library. The original code using pytorch_geometric can be found [here](https://github.com/neerajwagh/eeg-gcnn).

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: examples/pytorch/eeg-gcnn/.gitignore
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/.gitignore:1]: I guess we can simply remove this file.

File: examples/pytorch/eeg-gcnn/README.md
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/README.md:2]: "in here" -> "here"

File: examples/pytorch/eeg-gcnn/README.md
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/README.md:12]: - I think we can omit the item for "ArXiv Pre-print" and "Code" as they were covered in the section above.
- Can you add an explanation for the rest items so people can understand what they are and in what cases they can be helpful?

File: examples/pytorch/eeg-gcnn/README.md
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/README.md:25]: "datas" -> "datasets"

File: examples/pytorch/eeg-gcnn/README.md
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/README.md:5]: Is this one the same as the ArXiv version? If so, we can probably remove it and just mention it in the citation entry at the end.

File: examples/pytorch/eeg-gcnn/README.md
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/README.md:26]: Should `train.py` here be `main.py` instead?

File: examples/pytorch/eeg-gcnn/README.md
Comment (JOHNW02) [reply examples/pytorch/eeg-gcnn/README.md:5]: Hi Mufei,

This is a simplified version. The original code has precomputed patient indices. In this example, I removed the dependency on the indices.

File: examples/pytorch/eeg-gcnn/README.md
Comment (JOHNW02) [reply examples/pytorch/eeg-gcnn/README.md:26]: Yes! My bad.

Comment (mufeili) [reply examples/pytorch/eeg-gcnn/README.md]: Do you have default values for these arguments? If so, there's no need to add `--num_feats --num_nodes --gpu_idx --num_epochs --exp_name --batch_size`

File: examples/pytorch/eeg-gcnn/README.md
Comment (mufeili) [reply examples/pytorch/eeg-gcnn/README.md:5]: I see, thanks. Then perhaps you can mention explicitly that this is a simplified version.

File: examples/pytorch/eeg-gcnn/README.md
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/README.md:26]: There's no need to add `--num_feats --num_nodes --gpu_idx --num_epochs --exp_name --batch_size` as you have a default value for them.

File: examples/pytorch/eeg-gcnn/main.py
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/main.py:45]: I checked the preprocessed dataset [here](https://figshare.com/articles/software/EEG-GCNN_Supporting_Resources_for_Reproducibility/13251452), which does not seem to have the file `'norm_X'`

**Comment by mufeili** [comment]
Can you add an entry for this example in the indexing page [here](https://github.com/dmlc/dgl/tree/master/examples)?

File: examples/pytorch/eeg-gcnn/README.md
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/README.md:38]: How can a user reproduce the numbers here using `python main.py`?

File: examples/pytorch/eeg-gcnn/README.md
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/README.md:42]: How about 

emailed to John(_wei33@illinois.edu_)

You may also contact the authors:
- Neeraj: nwagh2@illinois.edu / [Website](http://neerajwagh.com/) / [Twitter](https://twitter.com/neeraj_wagh) / [Google Scholar](https://scholar.google.com/citations?hl=en&user=lCy5VsUAAAAJ)
- Yoga: varatha2@illinois.edu / [Website](https://sites.google.com/view/yoga-personal/home) / [Google Scholar](https://scholar.google.com/citations?user=XwL4dBgAAAAJ&hl=en)

File: examples/pytorch/eeg-gcnn/main.py
Comment (JOHNW02) [reply examples/pytorch/eeg-gcnn/main.py:45]: We haven't updated Figshare yet. We were dealing with a bug in the original code. So, some files need to be updated. We will do that next month, including norm_X. Also, norm_X is the normalized version of psd_featureds_data_X.

File: examples/pytorch/eeg-gcnn/README.md
Comment (JOHNW02) [reply examples/pytorch/eeg-gcnn/README.md:38]: I don't think they can. We are planning to update the original repo with both pytorch_geometric and dgl implementations, and release the trainning code as well. This should be done sometime next month too. For this example, can I just point people to the original repo for reproducing the stats?

File: examples/README.md
Comment (mufeili) [inline examples/README.md:19]: You also need to update the table at the top.

File: examples/pytorch/eeg-gcnn/main.py
Comment (mufeili) [reply examples/pytorch/eeg-gcnn/main.py:45]: In that case, we will need to wait for the files to be updated before we merge the PR.

File: examples/pytorch/eeg-gcnn/README.md
Comment (mufeili) [reply examples/pytorch/eeg-gcnn/README.md:37]: We expect a DGL example to be self-contained and able to reproduce the results reported in README. In that case, either we can wait till you get everything ready for this PR or we can simply close the PR and add a pointer to your own repo once you get everything ready there.

Comment (JOHNW02) [reply examples/pytorch/eeg-gcnn/README.md:37]: I see. Let me talk to Neeraj about this.

Comment (neerajwagh) [reply examples/pytorch/eeg-gcnn/README.md:37]: Hi @mufeili, maybe the following might clarify the design of the PR:

1) We built the PR with the following assumption: that people can go to https://github.com/neerajwagh/eeg-gcnn for reproducing published results on the original data while the DGL example here would enable others to bring their own datasets (X, y) and train our model using the data they care about. Compared to our original PyT implementation, here we have skipped: a) subject-level data splitting, b) subject-level aggregate evaluation, c) cross-validation. This makes the example a "simplified" version. **Nevertheless, we can update main.py and the results table so that main.py can reproduce the new README table (as long as users pull X, y from FigShare).** We can close the PR after that change while keeping a pointer to the PyT repo for the full/"complex" implementation. We'll elaborate on this simplification in the README too so users are aware of the difference between the two.

2) We can fix the absence of norm_X in FigShare by adding normalization after X is read in the code. (instead of updating the uploaded FigShare files).

Let me know if that works for you.

Comment (neerajwagh) [reply examples/pytorch/eeg-gcnn/README.md:37]: Meanwhile, the updates we make to our original PyT repo will remain independent of this example.

Comment (mufeili) [reply examples/pytorch/eeg-gcnn/README.md:37]: Hi @neerajwagh, the proposal sounds good to me in general. To be more specific, this example does not need to fully reproduce the experiments reported in your paper and I'm good as long as a user can run the script following the instructions in the README file and get similar numbers reported there.

Comment (neerajwagh) [reply examples/pytorch/eeg-gcnn/README.md:37]: @mufeili Sounds good! We'll make changes accordingly then. Thanks.

File: examples/pytorch/eeg-gcnn/README.md
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/README.md:8]: "in" -> "with"

File: examples/pytorch/eeg-gcnn/README.md
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/README.md:2]: Is it possible to provide a brief summary about the discrepancies between the original version and the simplified version?

File: examples/pytorch/eeg-gcnn/README.md
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/README.md:19]: Perhaps you want to mention that one can download with for example
```
wget https://ndownloader.figshare.com/files/25518170
```
and will need to unzip the downloaded file.

File: examples/pytorch/eeg-gcnn/README.md
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/README.md:20]: I don't know much about EEG data. My question is then do we need more detailed instructions on how to specify channels, electrodes and generate spectral coherence values and where should a user modify these? For the "where" part, can you add more detailed comments in the file?

File: examples/pytorch/eeg-gcnn/README.md
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/README.md:22]: Use `figshare_upload/psd_features_data_X` and `figshare_upload/labels_y` instead.

Comment (mufeili) [reply examples/pytorch/eeg-gcnn/README.md:22]: I saw `spec_coh_values.npy` and `standard_1010.tsv.txt` in both `psd_shallow_eeg-gcnn` and `psd_deep_eeg-gcnn`. Are they the same?

Comment (JOHNW02) [reply examples/pytorch/eeg-gcnn/README.md:22]: Yes, they are the same.

File: examples/pytorch/eeg-gcnn/README.md
Comment (JOHNW02) [reply examples/pytorch/eeg-gcnn/README.md:2]: There is a note under performance section. Do you want me to move it to the top?

Comment (mufeili) [reply examples/pytorch/eeg-gcnn/README.md]: I see. It looks good to me now.

File: examples/pytorch/eeg-gcnn/README.md
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/README.md:16]: Missing Pandas

File: examples/pytorch/eeg-gcnn/README.md
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/README.md:22]: Also need ```figshare_upload/master_metadata_index.csv```

File: examples/pytorch/eeg-gcnn/README.md
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/README.md:30]: Which experiment is this for? The shallow one or the deep one? How can a user reproduce the numbers reported below with this script?

File: examples/pytorch/eeg-gcnn/main.py
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/main.py:7]: `dump` not used

File: examples/pytorch/eeg-gcnn/main.py
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/main.py:27]: I guess we can have an argument to control the use of the deep variant or shallow variant.

File: examples/pytorch/eeg-gcnn/shallow_EEGGraphConvNet.py
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/shallow_EEGGraphConvNet.py:10]: `num_features_per_node` -> `num_features`

File: examples/pytorch/eeg-gcnn/shallow_EEGGraphConvNet.py
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/shallow_EEGGraphConvNet.py:22]: No need for `if type(x) == nn.Linear else None`

File: examples/pytorch/eeg-gcnn/shallow_EEGGraphConvNet.py
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/shallow_EEGGraphConvNet.py:23]: Ditto

File: examples/pytorch/eeg-gcnn/deep_EEGGraphConvNet.py
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/deep_EEGGraphConvNet.py:11]: Ditto

File: examples/pytorch/eeg-gcnn/deep_EEGGraphConvNet.py
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/deep_EEGGraphConvNet.py:29]: Ditto for `if type(x) == nn.Linear`

File: examples/pytorch/eeg-gcnn/deep_EEGGraphConvNet.py
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/deep_EEGGraphConvNet.py:35]: Remove L35

File: examples/pytorch/eeg-gcnn/deep_EEGGraphConvNet.py
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/deep_EEGGraphConvNet.py:38]: Remove L38

File: examples/pytorch/eeg-gcnn/deep_EEGGraphConvNet.py
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/deep_EEGGraphConvNet.py:43]: Ditto

File: examples/pytorch/eeg-gcnn/deep_EEGGraphConvNet.py
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/deep_EEGGraphConvNet.py:48]: Ditto

File: examples/pytorch/eeg-gcnn/deep_EEGGraphConvNet.py
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/deep_EEGGraphConvNet.py:53]: Ditto

File: examples/pytorch/eeg-gcnn/EEGGraphDataset.py
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/EEGGraphDataset.py:15]: Missing the descriptions for `indices` and `transform`

File: examples/pytorch/eeg-gcnn/EEGGraphDataset.py
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/EEGGraphDataset.py:23]: I'm a bit confused, is `x` edge weights or epochs?

File: examples/pytorch/eeg-gcnn/EEGGraphDataset.py
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/EEGGraphDataset.py:28]: Remove L28

File: examples/pytorch/eeg-gcnn/EEGGraphDataset.py
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/EEGGraphDataset.py:51]: L49-L51 are a bit confusing, either we remove them or we need to provide a better explanation.

File: examples/pytorch/eeg-gcnn/EEGGraphDataset.py
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/EEGGraphDataset.py:95]: remove L95

File: examples/pytorch/eeg-gcnn/main.py
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/main.py:84]: Remove this line

File: examples/pytorch/eeg-gcnn/README.md
Comment (JOHNW02) [reply examples/pytorch/eeg-gcnn/README.md]: I am asking Neeraj about spectral coherence values. I think it is a common thing with EEG. Not sure though.

Comment (mufeili) [reply examples/pytorch/eeg-gcnn/README.md]: I see. Thanks.

File: examples/pytorch/eeg-gcnn/README.md
Comment (mufeili) [inline examples/pytorch/eeg-gcnn/README.md:34]: I don't think you need `=` before `deep`.

**Comment by mufeili** [comment]
Thank you for the great job! This PR is almost ready to merge once we hear back from Neeraj and resolve the last conversation.

**Review by mufeili — APPROVED** [review]
LGTM

**Pull request merged**
